lib/planner/search-algorithms/offline-replanning.js

```diff
@@ -55,15 +66,57 @@ module.exports = offlineReplanning = {
         if (backtrackPlan !== null) {
           plan = backtrackPlan;
         } else if (!offlineReplanning._satisfiedActions.has(action)) {
+          newlySatisfiedActions.add(action);
           offlineReplanning._satisfiedActions.add(action);
           yield offlineReplanning.emitAsync('offlineReplanning.updateExistingPlansAndActionOccurrences', next);
+        } else {
+          let duplicate = offlineReplanning._comparePlan(plan);
+          while (duplicate !== undefined || offlineReplanning._isDuplicate(plan)) {
+            // console.log('duplicate of ' + duplicate);
+            // Backtrack
+            let {backtrackPlan, newSavedPlans, removed} =
+                yield offlineReplanning.emitAsync('offlineReplanning.backtrack',
+                    offlineReplanning._savedPlans.length - 2, // last good plan
+                    next);
+            if (backtrackPlan !== null) {
+              offlineReplanning._savedPlans = newSavedPlans;
+              if (configHandler.get('debug')) {
+                console.log(`Backtracked to: ${backtrackPlan.lastAction}`);
+              }
+              plan = backtrackPlan;
+
+              for (let removedAction of removed) {
+                if (newlySatisfiedActions.has(removedAction)) {
+                  newlySatisfiedActions.delete(removedAction);
+                  offlineReplanning._satisfiedActions.delete(removedAction);
+                }
+              }
+            } else {
+              throw new SimulatoError.PLANNER.FAILED_TO_BACKTRACK(
+                  'Unable to backtrack. No suitable backtrack point found'
+              );
+            }
+            duplicate = offlineReplanning._comparePlan(backtrackPlan);
+          }
         }
 
         if (backtrackPlan === null && configHandler.get('debug')) {
           console.log(`Added the following action to the plan: ${action}`);
         }
       }
-      offlineReplanning._savePlan(plan);
+
+      if (plan.path.length === 0) {
+        break;
+      } else if (looped && newlySatisfiedActions.size === 0) {
```

Comment:
Do we need this looped variables? Will `newlySatisfiedActions .size` ever be greater than 0 if it hasn't looped?

Reply:
I added "looped" to make this more easily testable. I needed some way to force it to skip that check.

Reply:
Sounds good